BinaryInteger advanced(by:) fixes

oscbyspro opened the pull request:
This patch fixes two unnecessary traps in BinaryInteger's advanced(by:) method. 

- `Int8.min.advanced(by: 128)`, etc.
- `UInt.max.advanced(by: Int.min)`, etc.

The first one is also mentioned in (https://github.com/apple/swift/issues/53747), w.r.t. `range.index(_:offsetBy:)`.

#### Before

```swift
if !Self.isSigned {
  return n < (0 as Int)
    ? self - Self(-n) // ------------------------ (1) crashes on UInt.max.advanced(by: Int.min)
    : self + Self(n)
}
if (self < (0 as Self)) == (n < (0 as Self)) {
  return self + Self(n) // ---------------------- (2) bit width comparison makes this unneccessary :tm:
}
return self.magnitude < n.magnitude
  ? Self(Int(self) + n)
  : self + Self(n) // --------------------------- (3) crashes on Int8.min.advanced(by: 128)
```

#### After

```swift
if Self.isSigned {
  return self.bitWidth < n.bitWidth // ---------- (2) constant-foldable when fixed-width (probably)
    ? Self(Int(truncatingIfNeeded: self) + n)
    : self + Self(truncatingIfNeeded: n) // ----- (3) can represent n because bit width is [>=]
} else {
  return n < (0 as Int)
    ? self - Self(UInt(bitPattern: ~n &+ 1)) // - (1) n.magnitude can represent all -n where n < 0
    : self + Self(UInt(bitPattern: n))
}
```

oscbyspro committed:
This patch fixes two unnecessary traps in BinaryInteger's advanced(to:) method.

- `Int8.min.advanced(by: 128)`, etc.
- `UInt.max.advanced(by: Int.min)`, etc.

oscbyspro committed:
Minor advanced(by:) enhancement.

glessard commented:
@swift-ci please test

glessard commented on test/stdlib/Integers.swift.gyb:
```diff
@@ -852,6 +852,9 @@ tests.test("Strideable") {
   expectEqual(dist(Int8.min, Int8.max), 255)
   expectEqual(dist(Int8.max, Int8.min), -255)
+  
+  expectEqual(Int8.min.advanced(by: 128), 0)
```
```suggestion
  expectEqual(Int8.min.advanced(by: Int(Int8.max)+1), 0)
```

oscbyspro committed:
Apply suggestions from code review

- Changed 128 integer literal to Int(Int8.max)+1 in unit tests (Int8.Stride is Int).

Co-authored-by: Guillaume Lessard <glessard@users.noreply.github.com>

glessard reviewed (approved):
This looks like a substantial code-gen improvement for the signed integer types, and a slight improvement for unsigned.

glessard commented:
@swift-ci please test
